File: termin-render/src/render_target_config.cpp
// render_target_config.cpp - RenderTargetConfig implementation
#include <termin/render_target_config.hpp>
#include <tcbase/tgfx_intern_string.h>
#include <cstring>

namespace termin {

tc_render_target_config RenderTargetConfig::to_c() const {
    tc_render_target_config c;
    tc_render_target_config_init(&c);

    c.name = name.empty() ? nullptr : tgfx_intern_string(name.c_str());
    c.kind = kind.empty() ? tgfx_intern_string("texture_2d") : tgfx_intern_string(kind.c_str());
    c.camera_uuid = camera_uuid.empty() ? nullptr : tgfx_intern_string(camera_uuid.c_str());
    c.xr_origin_uuid = xr_origin_uuid.empty() ? nullptr : tgfx_intern_string(xr_origin_uuid.c_str());
    c.width = width;
    c.height = height;
    c.dynamic_resolution = dynamic_resolution;
    c.color_format = color_format.empty() ? nullptr : tgfx_intern_string(color_format.c_str());
    c.depth_format = depth_format.empty() ? nullptr : tgfx_intern_string(depth_format.c_str());
    c.clear_color = clear_color;
    std::memcpy(c.clear_color_value, clear_color_value, sizeof(c.clear_color_value));
    c.clear_depth = clear_depth;
    c.clear_depth_value = clear_depth_value;
    c.pipeline_uuid = pipeline_uuid.empty() ? nullptr : tgfx_intern_string(pipeline_uuid.c_str());
    c.pipeline_name = pipeline_name.empty() ? nullptr : tgfx_intern_string(pipeline_name.c_str());
    c.layer_mask = layer_mask;
    c.enabled = enabled;
    if (!pipeline_params.empty()) {
        c.pipeline_params = tc_value_dict_new();
        for (const auto& [slot, value] : pipeline_params) {
            if (slot.empty() || value.empty()) continue;
            tc_value_dict_set(&c.pipeline_params, slot.c_str(), tc_value_string(value.c_str()));
        }
    }

    return c;
}

RenderTargetConfig RenderTargetConfig::from_c(const tc_render_target_config* c) {
    RenderTargetConfig cfg;
    if (!c) return cfg;

    cfg.name = c->name ? c->name : "";
    cfg.kind = c->kind ? c->kind : "texture_2d";
    cfg.camera_uuid = c->camera_uuid ? c->camera_uuid : "";
    cfg.xr_origin_uuid = c->xr_origin_uuid ? c->xr_origin_uuid : "";
    cfg.width = c->width;
    cfg.height = c->height;
    cfg.dynamic_resolution = c->dynamic_resolution;
    cfg.color_format = c->color_format ? c->color_format : "rgba16f";
    cfg.depth_format = c->depth_format ? c->depth_format : "depth32f";
    cfg.clear_color = c->clear_color;
    std::memcpy(cfg.clear_color_value, c->clear_color_value, sizeof(cfg.clear_color_value));
    cfg.clear_depth = c->clear_depth;
    cfg.clear_depth_value = c->clear_depth_value;
    cfg.pipeline_uuid = c->pipeline_uuid ? c->pipeline_uuid : "";
    cfg.pipeline_name = c->pipeline_name ? c->pipeline_name : "";
    cfg.layer_mask = c->layer_mask;
    cfg.enabled = c->enabled;
    if (c->pipeline_params.type == TC_VALUE_DICT) {
        for (size_t i = 0; i < tc_value_dict_size(&c->pipeline_params); i++) {
            const char* key = nullptr;
            tc_value* value = tc_value_dict_get_at(const_cast<tc_value*>(&c->pipeline_params), i, &key);
            if (!key || !key[0] || !value || value->type != TC_VALUE_STRING || !value->data.s) {
                continue;
            }
            cfg.pipeline_params[key] = value->data.s;
        }
    }

    return cfg;
}

} // namespace termin

File: termin-render/python/termin/render/render_target_config.py
"""RenderTargetConfig serialization helpers."""

from __future__ import annotations

from termin.render._render_native import RenderTargetConfig

__all__ = [
    "RenderTargetConfig",
    "serialize_render_target_config",
    "deserialize_render_target_config",
]


def serialize_render_target_config(config: RenderTargetConfig) -> dict:
    """Serialize RenderTargetConfig to a scene-dict representation."""
    result = {
        "name": config.name,
    }
    if config.kind and config.kind != "texture_2d":
        result["kind"] = config.kind
    if config.camera_uuid:
        result["camera_uuid"] = config.camera_uuid
    if config.xr_origin_uuid:
        result["xr_origin_uuid"] = config.xr_origin_uuid
    if config.dynamic_resolution:
        result["dynamic_resolution"] = True
    else:
        result["width"] = config.width
        result["height"] = config.height
    if config.color_format:
        result["color_format"] = config.color_format
    if config.depth_format:
        result["depth_format"] = config.depth_format
    if config.clear_color:
        result["clear_color"] = [float(v) for v in config.clear_color_value]
    if config.clear_depth:
        result["clear_depth"] = float(config.clear_depth_value)
    if config.pipeline_uuid:
        result["pipeline_uuid"] = config.pipeline_uuid
    if config.pipeline_name:
        result["pipeline_name"] = config.pipeline_name
    if config.layer_mask != 0xFFFFFFFFFFFFFFFF:
        result["layer_mask"] = hex(config.layer_mask)
    if not config.enabled:
        result["enabled"] = config.enabled
    if config.pipeline_params:
        result["pipeline_params"] = dict(config.pipeline_params)
    return result


def deserialize_render_target_config(data: dict) -> RenderTargetConfig:
    """Deserialize RenderTargetConfig from a scene-dict representation."""
    config = RenderTargetConfig()
    config.name = data.get("name", "")
    config.kind = data.get("kind", "texture_2d")
    config.camera_uuid = data.get("camera_uuid", "")
    config.xr_origin_uuid = data.get("xr_origin_uuid", "")
    config.width = data.get("width", 512)
    config.height = data.get("height", 512)
    config.dynamic_resolution = data.get("dynamic_resolution", False)
    config.color_format = data.get("color_format", "rgba16f")
    config.depth_format = data.get("depth_format", "depth32f")
    clear_color = data.get("clear_color", None)
    if isinstance(clear_color, (list, tuple)) and len(clear_color) >= 4:
        config.clear_color = True
        config.clear_color_value = tuple(float(v) for v in clear_color[:4])
    else:
        config.clear_color = False
        config.clear_color_value = (0.0, 0.0, 0.0, 1.0)
    clear_depth = data.get("clear_depth", None)
    if clear_depth is not None:
        config.clear_depth = True
        config.clear_depth_value = float(clear_depth)
    else:
        config.clear_depth = False
        config.clear_depth_value = 1.0
    config.pipeline_uuid = data.get("pipeline_uuid", "")
    config.pipeline_name = data.get("pipeline_name", "")

    layer_mask_raw = data.get("layer_mask", 0xFFFFFFFFFFFFFFFF)
    if isinstance(layer_mask_raw, str):
        config.layer_mask = int(layer_mask_raw, 16) if layer_mask_raw.startswith("0x") else int(layer_mask_raw)
    else:
        config.layer_mask = int(layer_mask_raw)

    config.enabled = data.get("enabled", True)
    pipeline_params = data.get("pipeline_params", {})
    if isinstance(pipeline_params, dict):
        config.pipeline_params = {
            str(slot): str(value)
            for slot, value in pipeline_params.items()
            if slot and value
        }
    return config
